### ISSUE TITLE
CV Container returns Invalid Container id error randomly

### ISSUE THREAD

**Issue by titom73** [aristanetworks/ansible-cvp#143]
Issue Type
------

 - Bug Report

Module Name
------

- _Module:_ `cv_container`
- _Ansible:_ `2.9.x`
- _arista.cvp:_ `>1.0.3a`
- _CloudVision server:_ `2019.1.2` / `2019.1.3`
- _Referenced issues:_ #142 / internal reports

Summary
------

When running `cv_container` tasks. Cloudvision server may return an error even if ansible tasks are correctly executed.

Steps to reproduce
------

One of playbook where issue was reported (#142).

```yaml
---
- name: Playbook to validate and demonstrate cv_container module.
  hosts: cvp
  connection: local
  gather_facts: no
  collections:
    - arista.cvp
  
  vars:
    containers_provision:
      IPFabric1:
        parent_container: Tenant
        devices:
        - Spine1
      IPFabric:
        parent_container: Tenant
        devices:
        - Leaf1
        - Leaf2

  tasks:
    - name: "Gather CVP facts from {{inventory_hostname}}"
      arista.cvp.cv_facts:
      register: cvp_facts

    - name: Display CVP info
      debug:
        msg: "{{cvp_facts.ansible_facts.cvp_info}}" 

    - name: Display CVP Facts
      debug:
        msg: "{{cvp_facts}}" 

    - name: "Build Container topology on {{inventory_hostname}}"
      arista.cvp.cv_container:
        topology: '{{containers_provision}}'
        cvp_facts: '{{cvp_facts.ansible_facts}}'
        #save_topology: false
        mode: override
      register: CVP_CONTAINERS_RESULT

    - name: Display tasks from cv_container
      debug:
        msg: "{{CVP_CONTAINERS_RESULT.data.tasks}}" 

    - name: 'Execute all pending tasks and wait for completion for 60 seconds'
      arista.cvp.cv_task:
        tasks: "{{ CVP_CONTAINERS_RESULT.data.tasks }}"
        wait: 60

    - name: "Refresh CVP facts from {{inventory_hostname}}"
      arista.cvp.cv_facts:
      register: cvp_facts

    - name: Display CVP Facts
      debug:
        msg: "{{cvp_facts}}" 
```

Expected results

```
[cvp_2019]: CHANGED
```

Actual results
------

```shell
fatal: [cvp_2019]: FAILED! => {"changed": false, "msg": 
"GET: https://x.x.x.x:443/web/provisioning/getContainerInfoById.do?containerId= : 
Request Error: Invalid Container id"}
```

**Comment by titom73** [comment]
Issue is applicable to 2 different use cases:

- Topology creation in override mode: tree was built on the wrong reversed list of container in `delete_unused_containers`
- Topology deletion with extra containers with attached devices: Tree topology and deletion process were based on wrong container list.

Fix is going to be available in a specific branch for testing (@roman-so)

**Comment by roman-so** [comment]
Hello Thomas,

Thanks for the update.
We still have exactly the same issue: 

> "msg": "GET: https://x.x.x..206:443/web/provisioning/getContainerInfoById.do?containerId= : Request Error: Invalid Container id"
> }

We rebuilt the image from the idicated branch, here are the logs:

> $ git log
> commit 2fab196a8fd381d2a004da2eef810af2f2108fa0 (HEAD -> issues/142-invalid-container-id, origin/issues/142-invalid-container-id)
> Author: Thomas Grimonet <tgrimonet@arista.com>
> Date:   Mon Mar 2 13:52:46 2020 +0100
> 
>     Fix deletion problem w/ additional containers #142
>    
>     Fix problem in delete_topology (state absent) and
>     delete_unused_containers (mode override) where reversed list where not
>     set to the correct tree list.

The playbook executed:
```
---
- name: Playbook to validate and demonstrate cv_container module.
  hosts: cvp
  connection: local
  gather_facts: no
  collections:
    - arista.cvp
  
  vars:
    containers_provision:
      IPFabric1:
        parent_container: Tenant
        devices:
        - arista_Spine1
      IPFabric:
        parent_container: Tenant
        devices:
        - arista_Leaf1
        - arista_Leaf2

  tasks:
    - name: "Gather CVP facts from {{inventory_hostname}}"
      arista.cvp.cv_facts:
      register: cvp_facts

    - name: Display CVP info
      debug:
        msg: "{{cvp_facts.ansible_facts.cvp_info}}" 

    - name: Display CVP Facts
      debug:
        msg: "{{cvp_facts}}" 

    - name: "Build Container topology on {{inventory_hostname}}"
      arista.cvp.cv_container:
        topology: '{{containers_provision}}'
        cvp_facts: '{{cvp_facts.ansible_facts}}'
        #save_topology: false
        mode: override
      register: CVP_CONTAINERS_RESULT

    - name: Display tasks from cv_container
      debug:
        msg: "{{CVP_CONTAINERS_RESULT.data.tasks}}" 

    - name: 'Execute all pending tasks and wait for completion for 60 seconds'
      arista.cvp.cv_task:
        tasks: "{{ CVP_CONTAINERS_RESULT.data.tasks }}"
        wait: 60

    - name: "Refresh CVP facts from {{inventory_hostname}}"
      arista.cvp.cv_facts:
      register: cvp_facts

    - name: Display CVP Facts
      debug:
        msg: "{{cvp_facts}}" 
```
Please find below the verbose output from the execution of the playbook
[output_issue_fix_test.txt](https://github.com/aristanetworks/ansible-cvp/files/4277368/output_issue_fix_test.txt)

After the execution of the playbook, In CVP we have the task created but not executed
![image](https://user-images.githubusercontent.com/59204786/75708868-e6de6700-5cc1-11ea-993b-9e844c4ecf50.png)

**Comment by titom73** [comment]
Hello @roman-so 

Regarding task generated and not executed is an expected behavior: task is generated by `cv_container` before starting to delete unused containers as stated by the `mode: merge` in your task. Because `cv_container` fails to delete a container, module do not generate correct output and ansible stop.

Based on output you shared, it seems `cv_container` try to delete a container with incorrect information. I have tested it in our lab and unfortunately I was not able to reproduce this behavior:

- 2 containers: One container with 2 devices and move one device to another container
- 3 containers: One container with 2 devices, move one device to another container, last container is deleted by ansible.
- 3 containers: One container with 2 devices, move one device to another container, last container (with one configlet attached) is deleted by ansible.
- 4 containers: One container with 2 devices, move one device to another container, one container with already one device attached and not desccribed in playbook, last container is deleted by ansible.

In the last commit, debug mode is activated meaning you should have a file named `cv_container.log` close to your playbook. Could you share this file as it will help us to get more information about what happens here ?

**Comment by AJNOURI** [comment]
We've tested "merge" successfully, but sill getting errors for "delete" and "override" every time.
Below the logs for each case:

[cv_container.merge.log](https://github.com/aristanetworks/ansible-cvp/files/4282870/cv_container.merge.log)
[cv_container.override.log](https://github.com/aristanetworks/ansible-cvp/files/4282871/cv_container.override.log)
[cv_container.delete.log](https://github.com/aristanetworks/ansible-cvp/files/4282872/cv_container.delete.log)

**Comment by titom73** [comment]
After reviewing logs, difference seems to be located in API calls within [`is_container_empty`](https://github.com/aristanetworks/ansible-cvp/blob/fe6dcd1b43b97cb5fea3365af829e0bbc64300bb/ansible_collections/arista/cvp/plugins/modules/cv_container.py#L545) and [`get_devices_in_container`](https://github.com/aristanetworks/ansible-cvp/blob/fe6dcd1b43b97cb5fea3365af829e0bbc64300bb/ansible_collections/arista/cvp/plugins/module_utils/cv_api2019.py#L918) of the API manager

__non-working scenario__

```
2020-03-02 18:55:06,504 * is_container_empty - get_devices_in_container IPFabric
2020-03-02 18:55:06,690 https://xxx.xxx.xxx.206:443 "GET /web/provisioning/searchTopology.do?queryParam=IPFabric&startIndex=0&endIndex=0 HTTP/1.1" 200 407
2020-03-02 18:55:06,766 https://xxx.xxx.xxx.206:443 "GET /web/inventory/devices?provisioned=IPFabric HTTP/1.1" 200 None
2020-03-02 18:55:06,834 https://xxx.xxx.xxx.206:443 "GET /web/provisioning/getContainerInfoById.do?containerId= HTTP/1.1" 200 60
2020-03-02 18:55:06,837 GET: https://xxx.xxx.xxx.206:443/web/provisioning/getContainerInfoById.do?containerId= : Request Error: Invalid Container id
```

__Working scenario__

```
2020-03-03 21:53:39,624 * is_container_empty - get_devices_in_container IPFabric1
2020-03-03 21:53:39,753 https://xxx.xxx.xxx.164:443 "GET /web/provisioning/searchTopology.do?queryParam=IPFabric1&startIndex=0&endIndex=0 HTTP/1.1" 200 257
2020-03-03 21:53:39,894 https://xxx.xxx.xxx.164:443 "GET /web/inventory/devices?provisioned=IPFabric1 HTTP/1.1" 200 None
2020-03-03 21:53:40,005 https://xxx.xxx.xxx.164:443 "GET /web/provisioning/getContainerInfoById.do?containerId=undefined_container HTTP/1.1" 200 185
2020-03-03 21:53:40,088 https://xxx.xxx.xxx.164:443 "GET /web/provisioning/getContainerInfoById.do?containerId=container_1532_840806557914744 HTTP/1.1" 200 170
```

In non-working scenario, we can see API call does not include `containerId` in the URL whereas it is part of the call in the working scenario.

**Comment by titom73** [comment]
Fixed and will be merged in PR #144
